pkg/servers/servers.go

```diff
@@ -54,21 +56,51 @@ func shutdown(name string, server *http.Server) {
 	}
 }
 
-func (s *ServerInfo) ListenAndServe() {
+func (s *ServerInfo) awaitConnection(address string, numberOfRetries int, timeout time.Duration, failureExpression string) error {
+	for i := 0 ; i < numberOfRetries; i++ {
+		_, err := net.DialTimeout("tcp", address, timeout)
+		if err == nil {
+			return nil
+		}
+		time.Sleep(timeout)
+	}
+	return fmt.Errorf(failureExpression, address, timeout)
+}
+
+func (s *ServerInfo) ListenAndServe() error {
+	numberOfRetries := 10000
+	timeout := 10 * time.Millisecond
 	if s.HTTP != nil {
 		go s.httpListen()
 	}
+
 	if s.HTTPS != nil {
 		go s.httpsListen()
 	}
+
+	if s.HTTP != nil {
+		err := s.awaitConnection(s.HTTP.Addr, numberOfRetries, timeout, "the HTTP server at %s did not start within the timeout of %d")
+		if err != nil {
+			return err
+		}
+	}
+
+	if s.HTTPS != nil {
+		err := s.awaitConnection(s.HTTPS.Addr, numberOfRetries, timeout, "the HTTPS server at %s did not start within the timeout of %d")
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
 func (s *ServerInfo) Shutdown() bool {
 	if s.HTTPS != nil {
-		shutdown("HTTPS", s.HTTPS)
+		go shutdown("HTTPS", s.HTTPS)
```

Comment:
No so much a fan of this. I'd prefer to somehow tell this function to shutdown more quickly just for unit tests.

But I also don't want to deal with changing the signature as that would mean a change to assisted-service as well